### DISCUSSION TITLE
problem with ruby file server - TCPServer Error: Address already in use

### DISCUSSION THREAD

**Discussion by MagnaFrisia90** [HCL-TECH-SOFTWARE/connections-automation#247] [Q&A] [answered]
Did anybody manage to setup ruby with the provided example command ?

I do always get the error message "TCPServer Error: Address already in use" eventhough there is no running process.
I have found 1 guy [on stackoverflow](https://stackoverflow.com/questions/60471315/ruby-webrick-think-every-local-port-is-already-in-use) who also has that same problem, but no answer provided : 

[root@solcon15 install]# ruby -run -ehttpd . -p8002
[2023-02-06 15:52:59] INFO  WEBrick 1.3.1
[2023-02-06 15:52:59] INFO  ruby 2.0.0 (2015-12-16) [x86_64-linux]
[2023-02-06 15:52:59] WARN  TCPServer Error: Address already in use - bind(2)

**Comment by umeli** [comment]
Hi
Did you check the open ports ? 
Does
netstat -tulpn | grep 8001
return anything ?

**Comment by stoeps13** [comment]
In my eyes it makes no sense to install ruby just to have an HTTP server serving installation files @sabrina-yee, Ansible and Docs need Python, so I would use:
`python -m http.server 8001`
started in the folder where the installation files are located.

**Comment by sabrina-yee** [comment]
As long as the installation kits are available to download via HTTP anything can be used.  If that server happens to be one of the deployment machines then yes python would work too.

@MagnaFrisia90 it does seem that the port is in use.  Please check per @umeli suggestion.   I see that you want to use port 8002 instead of 8001 so check:
`netstat -tulpn | grep 8002`

**Reply by MagnaFrisia90** [reply]
The problem is not with the ports used. This is a CentOS7 minimal image. Without any software/yum package installed. There can't be any application listening. I did all kind of ps/lsof/netstat commands prior creating the topic. Thats also why I linked the Stackoverflow-article:

> So I change the port number -p8000 to p8001 and I get the same error. I try 8002, 8003, 8888, 1313, 8004. I the same error on every single port number. Ruby, or Webrick, thinks that every port is already in use.
All the solutions to this problem I can find online suggest finding whatever process is hogging the port using commands like lsof | grep '8000' or lsof -wni tcp:8000 and then killing that process. But those commands don't return anything. There are no processes using those ports.

I think the problem is with the interface the httpd binds to. But I have no idea how to configure ruby httpd.

**Reply by sabrina-yee** [reply]
OK, will python work as Christoph suggested above?  Using ruby is not a requirement as long as the installation kit can be downloaded via http.

**Reply by MagnaFrisia90** [reply]
sadly I am also not able to get a python http server running.
As far as I understand I will need pip for Python to install the http.server module ? But this does not work on my CentOS7 image:

> [root@solcon13 cp]# pip install http.server
> Collecting http.server
>   Could not find a version that satisfies the requirement http.server (from versions: )
> No matching distribution found for http.server
> You are using pip version 8.1.2, however version 23.0 is available.
> You should consider upgrading via the 'pip install --upgrade pip' command.
> [root@solcon13 cp]# pip install --upgrade pip
> Collecting pip
>   Downloading https://files.pythonhosted.org/packages/b5/16/5e24bf63cff51dcc169f43bd43b86b005c49941e09cc3482a5b370db239e/pip-23.0.tar.gz (2.1MB)
>     100% |████████████████████████████████| 2.1MB 533kB/s
>     Complete output from command python setup.py egg_info:
>     Traceback (most recent call last):
>       File "<string>", line 1, in <module>
>       File "/tmp/pip-build-CPlqWN/pip/setup.py", line 7
>         def read(rel_path: str) -> str:
>                          ^
>     SyntaxError: invalid syntax
> 
>     ----------------------------------------
> Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-CPlqWN/pip/
> You are using pip version 8.1.2, however version 23.0 is available.
> You should consider upgrading via the 'pip install --upgrade pip' command.

Any further ideas ?

**Comment by stoeps13** [comment, marked as answer]
What Python version is installed? The error looks more like a compiler or run error, pip itself works.

Tons of options are available to run a local HTTP server: https://gist.github.com/willurd/5720255 You just need a service to provide the downloaded files to the machines.

**Reply by MagnaFrisia90** [reply]
2.7.5  

I will just try to run another HTTP-server

**Reply by stoeps13** [reply]
2.7.5 needs `python -m SimpleHTTPServer 8000`

**Reply by MagnaFrisia90** [reply]
This works for me ! Thanks
